Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,476 @@
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the request was applied.
+  optional int64 index = 4;
+  // term of raft when the request was applied.
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
+  optional bytes key = 1;
+  // if the end_key is not given, it is a get.
+  // if the end_key is given, it gets the keys in range [key, end_key).
+  optional bytes end_key = 2;
+  // limit the number of keys returned.
+  optional int64 limit = 3;
+  // if not all the keys are returned, a continue id will be 
+  // returned in the range response if consistent is set.
+  // the returned continue id only works with the previous 
+  // etcdserver the client connected to. 
+  optional bool consistent = 4
+  // continue the previous range request in a consistent
+  // manner if a continue id is provided.
+  optional int64 continue_id = 5;
+}
+
+message RangeResponse {
+  optional ResponseHeader header = 1;
+  repeated KeyValue kvs = 2;
+  optional continue_id = 3;
+  // the next range request with continue_id must
+  // be issued with in continue_timeout to the
+  // same server.
```

Comment:
> I still really don't like this limitation (which comes from choice of index implementation)

I do not quite like this either. Especially, it the client is talking to etcdserver over a proxy... This will be a big pain... But to be honest, I do not think consistent range query is a common use case...